Adding pellet printer suppor to OrcaSlicer

analysis230 opened the pull request:
Large format printers with print volumes in the order of 1m<sup>3</sup> generally use pellets for printing. The overall tech is very similar to FDM printing. It is FDM printing, but instead of filaments, it uses pellets. 

The difference here is that where filaments have a `filament_diameter` that is used to calculate the volume of filament ingested, pellets have a particular `flow_coefficient` that is empirically devised for that particular pellet. 

We are translating the `flow_coefficient` into `filament_diameter` so that everything works just like it does already with very minor adjustments. 

`filament_diameter = sqrt( (4 * flow_coefficient) / PI )`

`sqrt` just makes the relationship between `flow_coefficient` and volume linear.

This will allow a lot of accessibility for an increasing market of pellet 3d printers. Ginger additives, Colossus and CEAD being some of the leading names.

SoftFever commented on src/libslic3r/PresetBundle.cpp:
```diff
@@ -1533,6 +1533,27 @@ void PresetBundle::load_installed_filaments(AppConfig &config)
+								double pellet_flow_coefficient_value = filament->config.opt_float("pellet_flow_coefficient", 0);
+								double calculated_filament_diamter   = Preset::convert_pellet_flow_to_filament_diameter(
+									pellet_flow_coefficient_value);
+								filament->config.set_key_value("filament_diameter",
```
Since there is logic to handle the sync issue in Tab.cpp, is there any reason to have this check here? It's not a desired behavior here

SoftFever commented on src/libslic3r/PresetBundle.cpp:
```diff
@@ -1533,6 +1533,27 @@ void PresetBundle::load_installed_filaments(AppConfig &config)
+                            }
+                            else 
+                            {
+								double filament_diameter = filament->config.opt_float("filament_diameter", 0);
```
This will effectively change all existing normal printer's profile which don't use pellet_flow_coefficient.
So this code should be removed

analysis230 replied:
I will think of a workaround to make that not be an issue

analysis230 replied:
removed this code as this was unnecessary.

SoftFever commented:
Can I check what's the status of this PR?

analysis230 commented:
Some more changes are in the works. I have added a printers from the company that hired me to make the changes "Ginger Additive".

Here's our fork: https://github.com/gingeradditive/OrcaSlicer/tree/pellet-printer-implementation-rebased-onto-main

Just going to finalize things on Monday. and raise a proper pull request

SoftFever commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -1786,6 +1786,12 @@ void PrintConfigDef::init_fff_params()
 
+    def = this->add("pellet_flow_coefficient", coFloats);
+    def->label = L("Pellet flow coefficient");
+    def->tooltip = L("Pellet flow coefficient is emperically derived and allows for volume calculation for pellet printer");
```
Could you add more detailed descriptions here to help users understand how to calculate or set the parameter for their printer? Specifically, it would be helpful to include the relationship or formula that links this parameter to the filament diameter.

analysis230 replied:
Okay. I will add some helpful comments here and make it more coherent.

analysis230 replied:
Please take a look at the explanation. Hopefully this is clear enough.

SoftFever commented:
@analysis230 
Thanks.
I have two suggestions/feedbacks:
1. For the explanation, it's great to have a code comment👍 Could you also add the explanation especially the equation in the tooltip message as well? So that normal users could understand the parameter as well
2. When I test this PR, I noticed the Orca won't load the Ginger profiles.

analysis230 commented:
1. Added a tooltip with the formula. Also added a page to the Wiki so that normal users can take a look and understand better.
2. There was a minor problem with the json configs. Just checking the "3.0" variant was importing all configs. Fixed.

SoftFever reviewed (approved):
Looks good
Thanks

yw4z commented:
@vipulrajan hi, pellets.svg appears which place on UI?

vipulrajan commented:
It's used instead of the filament spool icon on the filament tab when you change to pellet settings.
